### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by s3theweb** [AdguardTeam/AdguardBrowserExtension#2594]
### AdGuard Extension version

4.2.209

### Browser version

Firefox 119.0.1

### OS version

Windows 11

### What filters do you have enabled?

AdGuard Base filter, Filter unblocking search ads and self-promotion, AdGuard Russian filter, AdGuard Chinese filter, AdGuard German filter, AdGuard Japanese filter, AdGuard Spanish/Portuguese filter, EasyList Italy, Liste AR, EasyList Thailand

### What Stealth Mode options do you have enabled?

Not enabled

### Issue Details

Over the course of about day or so most tabs will trend towards using a gigabyte or more of memory usage. It seems to be worse on pages where things are actually blocked, but even the most basic websites like example.com will use up hundreds of megabytes of memory when left open for a period of time. The google search homepage grows in memory usage almost 5 times faster than example.com
Disabling the adguard extension fixes it.

I found that it was Adguard specifically because a javascript memory snapshot on any tab showed hundreds of megabytes allocated to scripts within the extension.

### Expected Behavior

Basic pages should not use a gigabyte of memory.

### Screenshots

Screenshot 1:

![image](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/68665835/b21cd760-e96f-43e5-8e1b-2a724cd50064)
This is a javascript memory snapshot of example.com and moz-extension://3aeeed36-... is Adguard's extension path. Example.com has no javascript of its own.

### Additional Information

I am running Firefox and I also have uBlock origin, ghostery, tamper-monkey(which does not have any global scripts), and dark reader(not enabled for tabs in question) installed. This only started happening a few weeks ago and I had not installed or removed any extensions.

**Comment by user31415192** [comment]
From @gorhill of [uBlock Origin](//github.com/gorhill/ublock):

> Do **NOT** use uBO with any other content blocker. uBO [performs](https://www.debugbear.com/blog/chrome-extension-performance-2021#how-do-ad-blockers-and-privacy-tools-affect-browser-performance) as well as or better than most popular blockers. Other blockers can prevent uBO's privacy or anti-blocker-defusing features from working correctly.

So you got three blockers all running at once: uBlock Origin, Ghostery and AdGuard. From the quote above, we can deduce that 
it's a bad idea to have multiple blockers running concurrently, as they don't play well with each other; therefore, I strongly recommend keeping only **one** of those blockers, while the others must go. 

Also, for this same reason, this issue is **INVALID** and shall be closed ASAP.

**Comment by alexx7311** [comment]
@s3theweb Please check if the problem persists on the latest beta build for Firefox https://agrd.io/extension_firefox_beta

**Comment by s3theweb** [comment]
I agree with user31415192 that having multiple blockers like this is not ideal, but because this happens in the page's JavaScript theoretically it could be triggered by some script on a page if it does something similar to what the other extensions are doing.

I tested on the latest beta build and still got the same behavior. Example.com got up to over 2GB of ram usage over a few days. And the extension ID for the offending scripts is the one associated with the beta extension.
![image](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/68665835/47f50ac2-20e2-4116-a5e1-f12e6bb520e2)

**Comment by InQuize** [comment]
I can confirm, had memory leak issues and AdGuard is the only global adblock active.
To my surprise, there was exactly 24GB of RAM consumed by 'Extensions' in `about:processes`, haven't made screenshot then (was too quick to prevent next Firefox crash), but it should not be a problem to reproduce if no change was made..
All that RAM was immediately cleared as soon as I started to toggle extensions off-on starting with AdGuard.

Not sure if it goes up with the number of visited tabs or proportional to time they were active or both, but I utilized manual tab unloading recently, this way seems tame, will keep an eye on it.

**Comment by alexx7311** [comment]
@s3theweb Is the problem still relevant?

**Comment by InQuize** [comment]
My problem might be different in that RAM per tab is relatively low, all tab proccesses are ~1GB (e.g. group of youtube tabs) or less. The overall 'Extensions' RAM utilization is quite high, only growing at unknown rates and seems to freed the majority of it only upon reboot, restart or AdGuard extension off-on toggle, like on screenshots below:

![mem-usage](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/12504278/32bb0add-da9c-48a0-8dff-153a67da1535)
![mem-reset](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/12504278/61a69127-fa72-4955-b0be-37d3c4183a3e)
I have a bunch of tabs and windows across 6 virtual desktops and 2 monitors, and 2-4 weeks usual uptimes. This one was over 15d.

I monitored following figures over last weeks:
ext v4.2.209 - 24GB, Win begins to kill processes, preventive extension toggle off-on cleares all
ext v4.2.209 - 11GB, Firefox crash
ext v4.2.209 - 6GB, toggle off-on cleares all, switch to standalone from this thread
ext beta v4.2.240 - 9GB peak, 8GB at time of Firefox crash, switch to v4.2.240 release
ext beta v4.2.240 - preventive toggle off-on with screenshots above

When I restart AdGuard plugin, 'Extensions' consumption always drops to 100mb or less and drops down on a bottom of RAM sorted list, while there are no negative effects on already opened tabs or new ones, everything works as expected and looks like issue with garbage collection rather than memory leak.

**Comment by Tasiorsa** [comment]
I have exactly the same problem. Firefox eats 14 GB per 12 hours and is still growing. I only have 8 tabs open. Tested on release version and standalone. When I disable Adguard or change to uBlock everything looks fine.

**Comment by Smok07** [comment]
with adguard - 3GB
![Screenshot_4](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/15860522/056cbc9f-abef-4d1e-94e7-49cfc349ed76)
without adguard - 243MB
![Screenshot_5](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/15860522/ead40238-bd66-4613-acb7-251416eb68c8)

**Comment by Smok07** [comment]
4.5 GB after 12 hours of browsing
![Screenshot_8](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/15860522/4311f1dd-348e-4d34-821b-34b0a461bf6d)

**fixed by disabling "Privacy"** (only "AdGuard Tracking Protection filter" enabled), and "Language-specific" -> "AdGuard Russian filter"

I think we have a conflict between "Firefox Tracking Protection" and "AdGuard Tracking Protection"

**Comment by Tasiorsa** [comment]
Same scenario here 10-12GB per 12h. If I have fewer filters enabled, Firefox uses less memory for eg 4-6 GB per 12h

**Comment by alexx7311** [comment]
@InQuize @Smok07 @Tasiorsa Does the problem reproduce when disabling all extensions except AdGuard?

**Comment by Tasiorsa** [comment]
Yes, the problem occurs on a new profile with only one Adguard extension. Latest test Firefox 121.0.1 & Adguard 4.3.10.

**Comment by InQuize** [comment]
> Does the problem reproduce when disabling all extensions except AdGuard?

Finally had the opportunity to do just that.
On v4.3.13 in just two working days (of medium intensity web research) "Extensions" consumption jumped over 7GB.

Prior to that I was running uBlock Origin with all the usual unrelated extensions on and the same metric was floating around 500mb for the whole time of 40d uptime, had zero OOM issues.

![2d-1ext](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/12504278/6de9dc95-ad05-4253-81f5-7701978f4c64)

**Comment by alexx7311** [comment]
@InQuize @Smok07 @Tasiorsa Please check if the problem persists on the latest beta version for Firefox at https://github.com/AdguardTeam/AdguardBrowserExtension/releases/tag/v4.3.31-beta
We've fixed the possible problem with the memory leak there.

**Comment by Tasiorsa** [comment]
I've installed latest release https://github.com/AdguardTeam/AdguardBrowserExtension/releases/tag/v4.3.31
Testing in progress :-)

**Comment by Tasiorsa** [comment]
Memory consumption still increases, but much slower than in older versions. So the improvement is visible.

**Comment by InQuize** [comment]
I did not notice much difference. At 3GB right now.
Issue might be much worse in private mode (and certainly scales the most with the amount of sites visited and/or tabs opened), at least I believe that is how I got initial double digit figures of RAM consumption, but it is just a feeling, I haven't had an opportunity to reproduce it with similar amounts of tabs using private mode yet.

**Comment by Tasiorsa** [comment]
On Firefox 124.0.1 and Adguard 4.3.35, the problem of high RAM consumption still exists. About 12 GB RAM per 24 hours and constantly increasing

**Comment by tormozit** [comment]
I confirm. I am using Firefox + AdGuard lastest versions. Since at least begin of 2023 I notice constant memory leaks by AdGurad (visible in memory snapshot in developer tools). Last check made today on Firefox 125 + AdGuard 4.3.46 - I see no change in speed of leaks. So I still need to restart Firefox every day or more frequently.

**Comment by Revertron** [comment]
I think this bug may be responsible for high memory consumption:
https://github.com/AdguardTeam/AdguardBrowserExtension/issues/2792

**Comment by Smok07** [comment]
I can confirm that this is indeed a bug [#2792 ](url)

**Comment by Smok07** [comment]
after 3 days - 6.7gb
much slower, but still has a problem 
![Screenshot_2](https://github.com/AdguardTeam/AdguardBrowserExtension/assets/15860522/3e72043b-62f4-47d6-9371-15c1f6b4f421)

**Comment by gaveitatry2** [comment]
I just wanted to report that I had the same problem with AdGuard AdBlocker having an enormous memory leak on Mozilla Firefox (but not Google Chrome). I always had two tabs pinned (Discord and Amazon) and one open tab (offline website). When I restarted the browser, it would use up around 900 MB of memory. But as time passed, even if I did not use the browser at all, the memory usage would keep increasing. It would go up to 2500 MB. Then 4000-4500 MB. And just keep increasing throughout the day. I made it to point the restart my browser as often as possible, at least twice a day. Disabling this AdGuard Adblocker extension did the job. But anyway, this memory leak encouraged me to try a new browser so I'm switching from Mozilla Firefox to Google Chrome now. Hopefully this switch will be worth it, but I am thinking it will be.

**Comment by alexx7311** [comment]
Does the problem reproduce on the latest 4.4.22 version?

**Comment by Tasiorsa** [comment]
> Does the problem reproduce on the latest 4.4.22 version?

Windows 11 Pro
Firefox 130.0
Adguard 4.4.22 - fresh install
Still the same issue - screen after 10 hours - 8 bookmarks
![ffadg](https://github.com/user-attachments/assets/56afa858-855e-42d8-92f7-bf8f0fdab533)

**Comment by a-bhamjee** [comment]
In 24 hours with about 100 tabs open in Firefox, while running uBlock, memory usage is hovering at 6-7GBs. With this extension running (even though its disabled on every site, expect YT, UBlock is disabled on YT), memory usage is over 24 GB and extremely high CPU and Disk usage (because of paging). This extension has a huge memory leak, but no one has fixed it, yet they somehow charge money for a version of it.

**Comment by jimsinenomine** [comment]
Appears to be fixed in 5.0.170 (based on scenarios from #2972).  Will report back if experience regression.

**Comment by InQuize** [comment]
This is confusing.
4.4.48 is equivalent to 5.0.170 ?
Because on it I see no difference. Switched for a couple of days specifically to test and got 5GB "Extensions" in `about:processes` that droped to 748MB when I disabled AdGuard in `about:addons`, same as previous versions.

**Comment by asgh** [comment]
It's happening with Version 5.1.94 on Firefox 138.0.4

about:memory shows that it's enormous numbers (like 10,000 or more) of data:image/* allocated and never freed.

**Comment by ghost** [comment]
Why AG consumes right when opening the browser 150MB? uBO Lite doesn't consume anything when I open the browser (Edge 142 Stable)

<img width="1920" height="1020" alt="Image" src="https://github.com/user-attachments/assets/39063226-cb64-4fd7-8e5c-5f29549fab5a" />
<img width="1920" height="1020" alt="Image" src="https://github.com/user-attachments/assets/621a65eb-4668-4f95-ac0a-343eaa6a9e47" />

**Comment by domdfcoding** [comment]
I'm seeing the same as @asgh. After under 24 hours Adguard for Firefox uses around 11GB, mainly taken up by `data:image/x-icon;base64,...` that are several hundred megabytes each.  What are these images? I can't see a way to get Firefox to show the whole string, just the first 30 bytes or so.

Edit: I killed Firefox's WebExtensions process and it's come back using 450MB or so, and not a single `data:image/x-icon;base64,...`  showing in the memory viewer.

**Comment by asgh** [comment]
> Edit: I killed Firefox's WebExtensions process and it's come back using 450MB or so, and not a single `data:image/x-icon;base64,...` showing in the memory viewer.

Yah, I've started killing this process once or twice a day now to keep thing working. I used to think I had to close Firefox each night, but actually simply killing this process keeps it working well.

about:processes

And then the X near that process.

**Comment by darkgob** [comment]
> I'm seeing the same as [@asgh](https://github.com/asgh). After under 24 hours Adguard for Firefox uses around 11GB, mainly taken up by `data:image/x-icon;base64,...` that are several hundred megabytes each. What are these images? I can't see a way to get Firefox to show the whole string, just the first 30 bytes or so.
> 
> Edit: I killed Firefox's WebExtensions process and it's come back using 450MB or so, and not a single `data:image/x-icon;base64,...` showing in the memory viewer.

You can get the full string by using Dev Tools.  This is what I had:
- 2028 copies (!!) of a 1397 x 1328 transparent PNG. (134 MB)
- 624 copies of <img alt="Image" src="https://github.com/user-attachments/assets/ad0e6676-4efe-474f-937d-e1af451fe24e" />. (29 MB)
- 936 copies of <img alt="Image" src="https://github.com/user-attachments/assets/77fef43a-1229-4b55-8827-26b79d8f8312" />. (7 MB)

**Comment by scripthunter7** [comment]
Thank you for the reports and for your patience. Identifying memory leaks can sometimes be tricky, but we've likely found the root cause of the reported issue: Firefox handles favicons as data URLs, and these are being unnecessarily retained in memory. We're investigating further and will get back to you soon.

**Comment by darkgob** [comment]
> Thank you for the reports and for your patience. Identifying memory leaks can sometimes be tricky, but we've likely found the root cause of the reported issue: Firefox handles favicons as data URLs, and these are being unnecessarily retained in memory. We're investigating further and will get back to you soon.

Oh, I knew those icons seemed familiar!  It's the City of Heroes Homecoming forums and Stack Overflow, respectively.  I have tabs open for both.  The data for both appears to be incomplete though.  It was more obvious with the CoH icon (white star on a red circle), but now that I know the first is StackOverflow I can see that the icon is indeed also cut off.  Not sure if that's actually the case in memory though or just an artifact of Firefox's memory report.

What about that big transparent PNG?  That is by far the biggest offender.

**Comment by scripthunter7** [comment]
@darkgob Thanks for checking!

Technically, that large transparent PNG could also come from a favicon, Firefox allows sites to use very large images as favicons, and they can end up exposed to extensions as data: URLs. That said, without knowing the exact source, it's hard to be 100% certain.

If you're willing to share which sites you usually have open when this happens, that would help a lot with further debugging. If you'd prefer not to post that publicly, you can send the details directly to us at extensions@adguard.com, and we'll take a closer look.

Thanks again for the detailed investigation, it's very helpful.

**Comment by darkgob** [comment]
> [@darkgob](https://github.com/darkgob) Thanks for checking!
> 
> Technically, that large transparent PNG could also come from a favicon, Firefox allows sites to use very large images as favicons, and they can end up exposed to extensions as data: URLs. That said, without knowing the exact source, it's hard to be 100% certain.
> 
> If you're willing to share which sites you usually have open when this happens, that would help a lot with further debugging. If you'd prefer not to post that publicly, you can send the details directly to us at [extensions@adguard.com](mailto:extensions@adguard.com), and we'll take a closer look.
> 
> Thanks again for the detailed investigation, it's very helpful.

I have a lot of tabs currently open, but those were the only 3 data URLs being held in memory, and almost every tab has a favicon.  Two tabs don't display a favicon, but the structures of both (a website and a local HTML file) are relatively pretty bare bones, with neither explicitly defining a favicon.  (I also haven't looked at them in quite a while so nothing should have been held in memory for them.  Yes, I probably could just bookmark and close them, but...)  The transparent PNG is also the only one with a MIME type of image/png, the others are image/x-icon.

**Comment by kodiakhub** [comment]
Does this `Fixed` apply to that issue too, since the issue seems similar: https://github.com/AdguardTeam/AdguardBrowserExtension/issues/2524#issuecomment-2784765913
